implement-directus-auth-in-nextjs-14/index.md

```diff
@@ -0,0 +1,995 @@
+
+## Implementing Server Actions for Auth Components
+
+Server Actions are **asynchronous functions** that are executed on the server. They can be used in Server and Client Components to handle form submissions and data mutations in Next.js applications.
+
+To send user credentials to the `Directus` backend on the user register form submission, navigate to the `lib` directory, create `action.js` file and add the following:
+
+```javascript
+"use server";
+import { revalidatePath } from "next/cache";
+import { directusLogin } from "./auth";
+import {
+  createItem,
+  createUser,
+  deleteItem,
+  readItem,
+  readItems,
+  readMe,
+  updateItem,
+  uploadFiles,
+  withToken,
+} from "@directus/sdk";
+import directus from "@/lib/directus";
+import { redirect } from "next/navigation";
+import { AUTH_USER, COOKIE_NAME } from "@/constants";
+import { cookies } from "next/headers";
+
+export const register = async (previousState, formData) => {
+  try {
+    const { first_name, last_name, email, password } =
+      Object.fromEntries(formData);
+    const result = await directus.request(
+      createUser({
+        first_name,
+        last_name,
+        email,
+        password,
+        role: process.env.USER_ROLE,
+      })
+    );
+    return { message: "Account Created!", status: 201 };
+  } catch (e) {
+    console.log(e);
+    const code = e.errors[0].extensions.code;
+    if (code === "RECORD_NOT_UNIQUE") {
+      return { message: "This user already exist", status: 409 };
+    }
+    return {
+      message: "An unexpected error occurred, please try again",
+      status: 500,
+    };
+  }
+};
+
+export const login = async (prevState, formData) => {
+  const { email, password } = Object.fromEntries(formData);
+
+  try {
+    const user = await directusLogin({ email, password });
+    if (user) {
+      redirect("/");
+    }
+  } catch (err) {
+    console.log(err);
+    if (err.message.includes("Wrong credentials!")) {
+      return { error: "Invalid username or password" };
+    }
+    throw err;
+  }
+};
+
+export const handleLogout = async () => {
+  cookies().delete(AUTH_USER);
+  cookies().delete(COOKIE_NAME);
+};
+
+export const userSession = async () => {
+  const cookieStore = cookies();
+  let token = undefined;
+  const cookie = cookieStore.get(COOKIE_NAME);
+  if (cookie?.value) {
+    token = JSON.parse(cookie?.value);
+  }
+
+  const user = {};
+  if (token != undefined) {
+    user.accessToken = token.data.access_token;
+    user.refreshToken = token.data.refresh_token;
+    getAuthUser(token)
+  }
+  return user;
+};
+
+export const getAuthUser = async (token) => {
+  const userData = await directus.request(
+    withToken(
+      token.data.access_token,
+      readMe({
+        fields: ["*"],
+      })
+    )
+  );
+  return userData;
+};
+```
+
+These functions handle user logout, registration, login and users session using the Directus SDK.
+
+## Creating the Auth Pages
```

Comment:
This whole section is a bit jarring in how it's structured. Can we remove this H2 and have two H2s instead: "Implementing Registration" and "Implementing Login". Under each, see each to it's conclusion. It doesn't need to have subheadings, but structurally I would something like this

- Server action
- Form
- Page
- Test